Fix  part #161: Introduce image parser

veena14cs opened the pull request:
## Explanation
This is replica of [#201](https://github.com/oppia/oppia-android/pull/201). It contains ImageUrl Parser. 
UrlImageParser class encounters an <img> tag from the HTML contents, extracts image url from the "src" attribute and returns the image. It has testcases implemented.

BenHenning committed:
Introduce first pass interface for ExplorationProgressController.

BenHenning committed:
Fill in the stubbed logic for ExplorationProgressController. Still no
tests to verify correctness.

Also, added a method to facilitate notifying of DataProvider changes on
the UI thread.

BenHenning committed:
Fix lateinit issue in ExplorationProgressController due to wrongly ordered
initialization.

BenHenning committed:
Fix a variaty of issues in the exp progress controller, properly hook it up to
the data controller, and start adding tests.

BenHenning committed:
Merge branch 'develop' into introduce-exploration-progress-controller

Conflicts:
	domain/build.gradle
	domain/src/main/java/org/oppia/domain/exploration/ExplorationDataController.kt

Also, migrate the data controller to the retriever.

BenHenning committed:
Created a separate ExplorationRetriever, hooked up
AnswerClassificationController, and attempted to make
ExplorationProgressController thread-safe.

The thread-safety led to significant interface changes in the progress
controller, and led to discovering some issues with the mediator live data
approach to interop coroutines and LiveData. This locking mechanism will
need to change since the optimal solution requires resolving #90.

BenHenning committed:
Merge branch 'develop' into introduce-exploration-progress-controller

BenHenning committed:
Change the locking mechanism for ExplorationProgressController to work
with the current MediatorLiveData implementation (see #90 for more
context). Fix existing progress controller tests and add a few more. All
current progress controller tests are passing.

BenHenning committed:
Finish tests for ExplorationProgressController and add test classification
support for the second test exploration (about_oppia).

BenHenning committed:
Merge branch 'develop' into introduce-exploration-progress-controller

veena14cs committed:
craeted image parser

Commit:
ObservableViewModel

Commit:
Merge remote-tracking branch 'upstream/introduce-exploration-progress-controller' into introduce-load-exploration-part-1

Commit:
Load exploration and open ExplorationActivity

Commit:
Test case for load exploration

veena14cs committed:
fetch exploration

Commit:
Connection with StateFragment

veena14cs committed:
update

veena14cs committed:
update

veena14cs committed:
Merge branch 'introduce-load-exploration-part-1' into exploration-player-content-interaction

veena14cs committed:
working on fetching customization_args

veena14cs committed:
Update StateFragmentPresenter.kt

veena14cs committed:
resolved conflicts

veena14cs committed:
Separated this in its own PR

veena14cs committed:
Delete UrlImageParser.kt

Separated this in its own PR.

veena14cs committed:
Update StateFragmentPresenter.kt

veena14cs committed:
Update build.gradle

veena14cs committed:
Create HtmlParser.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
working on testcases

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
added testcases.

veena14cs committed:
Update AndroidManifest.xml

veena14cs committed:
Update InteractionAdapter.kt

veena14cs committed:
Update InteractionAdapter.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
corrected variable names.

veena14cs committed:
updated variables.

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Merge branch 'develop' into introduce-image-parser-replica

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Merge branch 'develop' into introduce-html-parser

veena14cs committed:
Delete InteractionAdapter.kt

veena14cs committed:
deteted unused xml files

veena14cs committed:
Merge branch 'introduce-html-parser' into introduce-image-parser-replica

veena14cs committed:
merged html code.

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Merge branch 'introduce-html-parser' into introduce-image-parser-replica

veena14cs committed:
Delete UrlImageParserTest.kt

veena14cs committed:
used dagger injection for UrlImageParser.

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
Merge branch 'introduce-html-parser' into introduce-image-parser-replica

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
updated tests

rt4914 commented on app/src/main/java/org/oppia/app/activity/HtmlParserTestActivty.kt:
```diff
@@ -1,22 +1,26 @@
 package org.oppia.app.activity
```
I would suggest to shift this HtmlParserTestActivity to `testing` package or some other package with implements `HtmlParser`

veena14cs replied:
moved it to package testing.

rt4914 commented on app/src/main/java/org/oppia/app/activity/HtmlParserTestActivty.kt:
```diff
@@ -1,22 +1,26 @@
+    activityComponent.inject(this)
     setContentView(R.layout.activity_test_html_parser)
 
     var testHtmlContentTextView = findViewById(R.id.test_html_content_text_view) as TextView
```
Prefer writing it this way

`val testHtmlContentTextView: TextView = findViewById(R.id.test_html_content_text_view)`


I suggest try to solve warnings suggested by Android Studio.

veena14cs replied:
Okay will do that.

rt4914 commented on app/src/main/java/org/oppia/app/activity/HtmlParserTestActivty.kt:
```diff
@@ -1,22 +1,26 @@
     setContentView(R.layout.activity_test_html_parser)
 
     var testHtmlContentTextView = findViewById(R.id.test_html_content_text_view) as TextView
     var rawDummyString = "\u003cp\u003e\"Let's try one last question,\" said Mr. Baker. \"Here's a pineapple cake cut into pieces.\"\u003c/p\u003e\u003coppia-noninteractive-image alt-with-value=\"\u0026amp;quot;Pineapple cake with 7/9 having cherries.\u0026amp;quot;\" caption-with-value=\"\u0026amp;quot;\u0026amp;quot;\" filepath-with-value=\"\u0026amp;quot;pineapple_cake_height_479_width_480.png ...
```
Replace `var` with `val`

veena14cs replied:
corrected.

rt4914 commented on app/src/main/java/org/oppia/app/activity/HtmlParserTestActivty.kt:
```diff
@@ -1,22 +1,26 @@
 
 /** This is a dummy activity to test Html parsing. */
-class HtmlParserTestActivty : AppCompatActivity() {
+class HtmlParserTestActivty : InjectableAppCompatActivity() {
```
Typo in file name
`HtmlParserTestActivity`

veena14cs replied:
corrected.

rt4914 commented on utility/build.gradle:
```diff
@@ -43,7 +43,8 @@ dependencies {
       'com.google.dagger:dagger:2.24',
-      "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+      "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version",
+      'com.github.bumptech.glide:glide:4.9.0',
```
Prefer maintaining the alphabetical order.

veena14cs replied:
corrected.

rt4914 replied:
Still not in alphabetical order.

veena14cs replied:
Changed to alphabetical order.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/HtmlParser.kt:
```diff
@@ -11,7 +11,8 @@ private const val CUSTOM_IMG_FILE_PATH_ATTRIBUTE = "filepath-with-value"
 
 /** Html Parser for android TextView to parse Html tag. */
-class HtmlParser(private val context: Context, private val entityType: String, private val entityId: String) {
+class HtmlParser(private val context: Context, private val entityType: String, private val entityId: String,
```
Optional Nit: This can be made more readable by bringing each object to next line.

BenHenning replied:
This is also part of https://google.github.io/styleguide/javaguide.html#s4.5.1-line-wrapping-where-to-break (breaking at a higher syntactic level).

Here, you broke the line mid-way through the argument list, but the higher syntax is the beginning of the argument list. Thus, we should break at the ``(`` first, and then if all arguments don't fit then we should break along every argument at the ``,``.

veena14cs replied:
Fixed it.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/HtmlParser.kt:
```diff
@@ -29,11 +30,12 @@ class HtmlParser(private val context: Context, private val entityType: String, p
       );
```
Remove `;`

rt4914 replied:
Remove `;`

in 

`htmlContent = htmlContent.replace(CUSTOM_IMG_TAG, REPLACE_IMG_TAG, /* ignoreCase= */false);`

veena14cs replied:
done.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/ImageLoader.java:
```diff
@@ -0,0 +1,17 @@
+import com.bumptech.glide.Glide;
+import com.bumptech.glide.request.target.SimpleTarget;
+
+public interface ImageLoader {
```
Can we convert this java file to kotlin?

BenHenning replied:
+1

No Java files please. :) Unless we need to, in which case let's chat about it.

veena14cs replied:
Fixed it :)

rt4914 commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+
+  lateinit var htmlContentTextView: TextView
+  lateinit var context: Context
+  lateinit var entityType: String
```
Can be made `private`

BenHenning replied:
+1 Please make all values private here since none need to be exposed outside of this class.

veena14cs replied:
done.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+  lateinit var htmlContentTextView: TextView
+  lateinit var context: Context
+  lateinit var entityType: String
+  lateinit var entityId: String
```
Can be made `private`

veena14cs replied:
done.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  /***
+   * This method is called when the HTML parser encounters an <img> tag
```
`.` missing at end of sentence.

veena14cs replied:
done.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+   * @return Drawable : Drawable representation of the image
+   */
+  override fun getDrawable(urlString: String): Drawable {
+
```
Remove this empty line

veena14cs replied:
removed empty lines.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+import javax.inject.Inject
+import javax.inject.Singleton
+
+// TODO (#169) : Replace this with exploration asset downloader
```
`// TODO (#169): Replace this with exploration asset downloader`

BenHenning replied:
Actually:

``// TODO(#169): Replace this with exploration asset downloader.``

(The 'TODO' and '(' as well as the ')' and ':' should be close together).

veena14cs replied:
corrected.

rt4914 reviewed (changes requested):
Generally LGTM, just some nit changes.

@BenHenning I suggest you check the test-case approach.

veena14cs committed:
updated test

veena14cs committed:
moved to testing package

veena14cs committed:
Merge branch 'introduce-html-parser' into introduce-image-parser-replica

veena14cs committed:
Update HtmlParserTestActivty.kt

BenHenning commented on app/src/main/java/org/oppia/app/testing/HtmlParserTestActivty.kt:
```diff
@@ -1,22 +1,27 @@
-    val testHtmlContentTextView = findViewById(R.id.test_html_content_text_view) as TextView
-    val rawDummyString = "\u003cp\u003e\"Let's try one last question,\" said Mr. Baker. \"Here's a pineapple cake cut into pieces.\"\u003c/p\u003e\u003coppia-noninteractive-image alt-with-value=\"\u0026amp;quot;Pineapple cake with 7/9 having cherries.\u0026amp;quot;\" caption-with-value=\"\u0026amp;quot;\u0026amp;quot;\" filepath-with-value=\"\u0026amp;quot;pineapple_cake_height_479_width_480.png ...
-    val htmlResult: Spannable = HtmlParser(applicationContext, /* entityType= */"", /* entityId= */"").parseOppiaHtml(
+    var testHtmlContentTextView = findViewById(R.id.test_html_content_text_view) as TextView
```
Prefer this & next variable to be ``val``s if they don't need to be overwritten again.

veena14cs replied:
corrected.

BenHenning commented on app/src/main/java/org/oppia/app/testing/HtmlParserTestActivty.kt:
```diff
@@ -1,22 +1,27 @@
-    val htmlResult: Spannable = HtmlParser(applicationContext, /* entityType= */"", /* entityId= */"").parseOppiaHtml(
+    var testHtmlContentTextView = findViewById(R.id.test_html_content_text_view) as TextView
+    var rawDummyString = "\u003cp\u003e\"Let's try one last question,\" said Mr. Baker. \"Here's a pineapple cake cut into pieces.\"\u003c/p\u003e\u003coppia-noninteractive-image alt-with-value=\"\u0026amp;quot;Pineapple cake with 7/9 having cherries.\u0026amp;quot;\" caption-with-value=\"\u0026amp;quot;\u0026amp;quot;\" filepath-with-value=\"\u0026amp;quot;pineapple_cake_height_479_width_480.png ...
+    val htmlResult: Spannable = HtmlParser(applicationContext, /* entityType= */"exploration", /* entityId= */"DIWZiVgs0km-",urlImageParser).parseOppiaHtml(
```
Please pick an entity ID that doesn't correspond to a real exploration on oppia.org. We shouldn't be connecting to a network in our tests, anyway, so the value we use here shouldn't matter.

veena14cs replied:
Added this for testing. To test whether actual image is getting loaded. But anyway these testcases are going to change as per https://github.com/oppia/oppia-android/pull/201#discussion_r331313707. 
@BenHenning I am not clear on this so these test cases are pending.

rt4914 replied:
@veena14cs I think here @BenHenning is suggesting to use some other `entity-id` and do not use `DIWZiVgs0km-` because it actually exists in oppia.org

veena14cs replied:
ok corrected.

veena14cs committed:
Update HtmlParserTestActivty.kt

BenHenning commented on app/src/sharedTest/java/org/oppia/app/activity/HtmlParserTestActivty.kt:
```diff
@@ -50,8 +52,8 @@ class HtmlParserTestActivityTest {
     val htmlResult: Spannable = HtmlParser(
-      ApplicationProvider.getApplicationContext(), "", ""
-    ).parseOppiaHtml(
+      ApplicationProvider.getApplicationContext(), /* entityType= */ "",  /* entityId= */"",
```
Nit: space between ``*/`` and ``""``. Ditto elsewhere.

veena14cs replied:
corrected.

veena14cs committed:
fix issues

veena14cs committed:
fix issues.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/HtmlParser.kt:
```diff
@@ -11,7 +11,8 @@ private const val CUSTOM_IMG_FILE_PATH_ATTRIBUTE = "filepath-with-value"
 
 /** Html Parser to parse custom Oppia tags with Android-compatible versions. */
-class HtmlParser(private val context: Context, private val entityType: String, private val entityId: String) {
+class HtmlParser(private val context: Context, private val entityType: String, private val entityId: String,
```
Since Context and UrlImageParser could both be injected, I suggest using a factory for this parser and making it's constructor private. Changing it would look something like:

```kotlin
class HtmlParser private constructor(
  private val context: Context,
  private val urlImageParser: UrlImageParser,
  private val entityType: String,
  private val entityId: String
) {
  ...

  class Factory @Inject constructor(
    @ApplicationContext private val context: Context, private val urlImageParser: UrlImageParser
  ) {
    fun create(entityType: String, entityId: String): HtmlParser {
      return HtmlParser(context, urlImageParser, entityType, entityId)
    }
  }
}
```

Then using it to create a new parser:

```kotlin
@Inject HtmlParser.Factory htmlParserFactory
...

  val htmlParser = htmlParserFactory.create(entityType, entityId)
```

This allows you to still benefit from dependency injection but also pass in unbound values only known at runtime (like entityType/entityId).

veena14cs replied:
done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/HtmlParser.kt:
```diff
@@ -29,11 +30,12 @@ class HtmlParser(private val context: Context, private val entityType: String, p
+    val imageGetter = urlImageParser
     return if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
-      Html.fromHtml(htmlContent, Html.FROM_HTML_MODE_LEGACY, /* imageGetter= */null, /* tagHandler= */null) as Spannable
+      Html.fromHtml(htmlContent, Html.FROM_HTML_MODE_LEGACY, /* imageGetter= */imageGetter, /* tagHandler= */null) as Spannable
```
Super nit: the ``/* imageGetter= */`` isn't needed if you're passing in a variable, since the variable name is documenting the value being passed in. We only add argument comments for literals to provide context to what the value is corresponding to.

veena14cs replied:
Corrected.

veena14cs committed:
fixed issues

BenHenning commented on utility/src/main/java/org/oppia/util/parser/ImageParsingAnnotations.kt:
```diff
@@ -0,0 +1,12 @@
+import javax.inject.Qualifier
+
+/** Corresponds to the default GCS Resource Bucket Name. */
+@Qualifier annotation class DefaultGCSResource
```
Nit: ``DefaultGcsResource``

See: https://google.github.io/styleguide/javaguide.html#s5.3-camel-case.

veena14cs replied:
Corrected.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/ImageParsingModule.kt:
```diff
@@ -0,0 +1,30 @@
+  @DefaultGCSResource
+  @Singleton
+  fun provideDefaultGcsResource(): String {
+    return  "oppiaserver-resources/"
```
Nit: one space between ``return`` and ``"oppiaserver-resources/"``

veena14cs replied:
corrected.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/ImageLoader.java:
```diff
@@ -0,0 +1,17 @@
+
+public interface ImageLoader {
+
+  static void load(Context context, String path, int placeholder, SimpleTarget<Bitmap> target) {
```
Doesn't this need to be an instance method? I was expecting that we need to mock out ImageLoader in tests to avoid using Glide in the actual test environment. That means we'd have something like:
- ``interface ImageLoader``
- ``class GlideImageLoader implements ImageLoader``
- Tests bind a Mockito mock of ``ImageLoader`` that performs the load operation
- Places that need ``ImageLoader`` inject it from Dagger
- We need a module to bind ``GlideImageLoader`` to ``ImageLoader`` so that injecting ``ImageLoader`` in production code corresponds to the glide image loader

Note that we don't need to pass ``Context`` in since subclasses of ``ImageLoader`` will be injectable and can inject the application context.

rt4914 replied:
@veena14cs This comment has not been addressed.

veena14cs replied:
I have made changes to the implementation part regarding ImageLoader. Testcases are still pending.

veena14cs replied:
Filed an issue #277 for this case.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/ImageLoader.java:
```diff
@@ -0,0 +1,17 @@
+import com.bumptech.glide.Glide;
+import com.bumptech.glide.request.target.SimpleTarget;
+
+public interface ImageLoader {
```
Please add documentation.

rt4914 replied:
@veena14cs This comment has not been addressed

veena14cs replied:
added doc comment.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/ImageParsingAnnotations.kt:
```diff
@@ -0,0 +1,12 @@
+@Qualifier annotation class DefaultGcsPrefix
+
+/** Corresponds to the default Image Download Url Template. */
+@Qualifier annotation class ImageDownloadUrlTemplate
```
I think it's definitely worth documenting that strings bound to this template will have 3 format variables in them, and what those variables should correspond to.

rt4914 replied:
@veena14cs This conversation ha not been addressed.

veena14cs replied:
added doc comment.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+      gcsPrefix + gcsResource + imageDownloadUrlTemplate,
+      R.drawable.abc_ab_share_pack_mtrl_alpha,
+      target
+    );
```
Remove ``;``

veena14cs replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+  inner class UrlDrawable : BitmapDrawable() {
+    var drawable: Drawable? = null
+    override fun draw(canvas: Canvas) {
+      if (drawable != null)
```
Optional: could be made cleaner to avoid the ``!!``:

```kotlin
val currentDrawable = drawable
if (currentDrawable != null) {
  currentDrawable.draw(canvas)
}
```

This works because you're capturing the current drawable state into a ``val``, so it can't change after the null check (thus Kotlin can infer it's non-nullability, so no ``!!`` operator is necessary).

veena14cs replied:
done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+    }
+  }
+
+  inner class UrlDrawable : BitmapDrawable() {
```
Here & above: please make private since these classes do not need to be accessed outside of the outer class.

veena14cs replied:
made it private class.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  inner class UrlDrawable : BitmapDrawable() {
+    var drawable: Drawable? = null
```
Please make private. Ditto for other inner class parameters elsewhere in this PR.

veena14cs replied:
done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+    }
+  }
+
+  inner class UrlDrawable : BitmapDrawable() {
```
Does this drawable actually need to be an inner class, or can it be a regular class? I don't see any outer class members being accessed here.

veena14cs replied:
to access fields of UrlImageParser I made it inner class.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+
+  inner class BitmapTarget(private val urlDrawable: UrlDrawable) : SimpleTarget<Bitmap>() {
+    internal var drawable: Drawable? = null
+    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
```
Here & elsewhere: please put spacing between class fields and methods.

veena14cs replied:
done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+    ImageLoader.load(
+      context,
+      gcsPrefix + gcsResource + imageDownloadUrlTemplate,
+      R.drawable.abc_ab_share_pack_mtrl_alpha,
```
What is this?

veena14cs replied:
I took some default image for debugging.

rt4914 replied:
I think this should be removed because we don't want to merge this in `develop`

veena14cs replied:
removed default place holder.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,81 @@
+  lateinit var entityType: String
+  lateinit var entityId: String
+
+  fun setImageData(htmlContentTextView: TextView, context: Context, entityType: String, entityId: String) {
```
Suggest using a factory to set this via constructor (but still keep the necessary injections) per what I suggested above for the HtmlParser class. That's less fragile than relying on an external initialization method that might be forgotten to be called which would lead to unexpected results.

veena14cs replied:
Done as per suggestion, introduced factory class.

BenHenning reviewed (changes requested):
Thanks @veena14cs. Mostly nits, though I had a few design suggestions to hopefully tighten construction of HtmlParser and UrlImageParser, but also still retain DI capabilities.

I don't see any new tests for UrlImageParser--am I missing something? @rt4914 you specifically asked me to review test cases here, but I don't see any new tests being added for the URL cases.

veena14cs committed:
fixed issues.

veena14cs committed:
Update activity_test_html_parser.xml

veena14cs committed:
fixes

veena14cs committed:
fixes

veena14cs committed:
Update HtmlParserTest.kt

veena14cs committed:
Update HtmlParserTest.kt

veena14cs committed:
fixed issues.

veena14cs committed:
fixing issues.

veena14cs committed:
made use of factory

veena14cs committed:
Update HtmlParserTest.kt

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update AndroidManifest.xml

veena14cs committed:
Update HtmlParserTestActivity.kt

veena14cs committed:
nit

veena14cs committed:
Update AudioFragmentTest.kt

veena14cs committed:
updated

veena14cs committed:
Merge branch 'introduce-image-parser-replica' of https://github.com/oppia/oppia-android into introduce-image-parser-replica

veena14cs committed:
fixes

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Not used in this.

veena14cs committed:
fixes

veena14cs committed:
Merge branch 'introduce-image-parser-replica' of https://github.com/oppia/oppia-android into introduce-image-parser-replica

veena14cs committed:
Merge branch 'introduce-html-parser' into introduce-image-parser-replica

veena14cs committed:
Not used.

veena14cs committed:
Update CellularDataDialogFragmentTest.kt

veena14cs committed:
Merge branch 'develop' into introduce-image-parser-replica

veena14cs committed:
reverted

veena14cs committed:
Update CellularDataDialogFragmentTest.kt

veena14cs committed:
fixes

veena14cs committed:
Update ImageLoader.kt

veena14cs committed:
Update HtmlParser.kt

rt4914 commented:
> Thanks @veena14cs. Mostly nits, though I had a few design suggestions to hopefully tighten construction of HtmlParser and UrlImageParser, but also still retain DI capabilities.
> 
> I don't see any new tests for UrlImageParser--am I missing something? @rt4914 you specifically asked me to review test cases here, but I don't see any new tests being added for the URL cases.

@veena14cs did mention me to ask you to check for test-cases, so I communicated that to you.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/HtmlParser.kt:
```diff
@@ -0,0 +1,54 @@
+
+/** Html Parser to parse custom Oppia tags with Android-compatible versions. */
+class HtmlParser private constructor(
+  private val context: Context,
```
@veena14cs You are not using this context

veena14cs replied:
removed unused context.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/HtmlParser.kt:
```diff
@@ -0,0 +1,54 @@
+      htmlContent = htmlContent.replace("&amp;quot;", "")
+    }
+
+    val imageGetter =  urlImageParserFactory.create(htmlContentTextView, entityType, entityId)
```
Should it be `urlImageParserFactory ` or `htmlParserFactory `

veena14cs replied:
It has to be urlImageParserFactory to parse image from html tags.

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Update build.gradle

veena14cs committed:
fixed issues.

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update ImageParsingAnnotations.kt

veena14cs committed:
working on image loader

veena14cs committed:
Injected ImageLoader

veena14cs committed:
updated image loader.

veena14cs committed:
Update ActivityComponent.kt

veena14cs committed:
Update HtmlParserTestActivity.kt

rt4914 reviewed (approved):
LGTM

BenHenning commented on utility/src/main/java/org/oppia/util/parser/GlideImageLoader.kt:
```diff
@@ -0,0 +1,17 @@
+import com.bumptech.glide.request.target.SimpleTarget
+import javax.inject.Inject
+
+class GlideImageLoader @Inject constructor() : ImageLoader {
```
Please provide a brief KDoc for this class. E.g. something like the following is sufficient:

```kotlin
/** An [ImageLoader] that uses Glide. */
```

veena14cs replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/HtmlParsingAnnotaions.kt:
```diff
@@ -0,0 +1,6 @@
+import javax.inject.Qualifier
+
+/** Qualifier for injecting the application context. */
+@Qualifier annotation class ApplicationContext
```
Let's not use this. I'd rather not duplicate the annotations.

You can just inject Context directly and it will be the application context. Once we resolve #59 we'll be able to fix this fully (there should already be TODO tracking this in the application component).

veena14cs replied:
removed.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/HtmlParsingModule.kt:
```diff
@@ -0,0 +1,26 @@
+    return application
+  }
+
+  // TODO(#59): Remove this provider once all modules have access to the @ApplicationContext qualifier.
```
Similarly, you should be able to just inject Context directly into HtmlParser. Please remove this module since it shouldn't be necessary.

veena14cs replied:
removed.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/ImageLoader.kt:
```diff
@@ -0,0 +1,13 @@
+import com.bumptech.glide.request.target.SimpleTarget
+import androidx.annotation.DrawableRes
+
+/** Helper that handles loading and caching images from remote URLs.*/
```
This isn't really a helper, it's more of an interface describing what an image loader should be able to do. Consider:

```kotlin
/** A loader for downloading and caching images from remote URLs. */
```

veena14cs replied:
corrected.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/ImageLoader.kt:
```diff
@@ -0,0 +1,13 @@
+/** Helper that handles loading and caching images from remote URLs.*/
+interface ImageLoader {
+
+  fun load(context: Context, path: String, target: SimpleTarget<Bitmap>)
```
This should document what implementations should do, and what each parameter is for, e.g.:

```kotlin
/** Loads an image from the provided URL into the specified target, optionally caching it. */
```

Can we avoid passing in Context? We should be able to inject that into implementations. Also, should ``path`` be ``imageUrl`` or something similar? 'path' feels local, but the class KDoc indicates this is for downloading remote images.

As an aside, it'd be nice if we could abstract away from ``SimpleTarget`` to fully abstract away Glide, but the current implementation is reasonable. It does help make it easier to migrate away from Glide in the future if we would ever want to.

veena14cs replied:
Corrected KDoc comment,  change `path` to `imageUrl`. Removed passing of context and injected in implementation.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/ImageLoaderAnnotation.kt:
```diff
@@ -0,0 +1,7 @@
+
+/** Corresponds to image loader. */
+@Qualifier
+annotation class ImageLoaderAnnotation
```
Do we actually need this annotation? We should just be able to inject ``ImageLoader`` directly. Please remove, or follow up with me on why we need this.

veena14cs replied:
If I remove and directly inject ImageLoader its throwing me error 
` error: cannot access SimpleTarget
  class file for com.bumptech.glide.request.target.SimpleTarget not found`

veena14cs replied:
This is been resolved.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/ImageLoaderModule.kt:
```diff
@@ -0,0 +1,17 @@
+
+/** Provides image loading dependencies. */
+@Module
+class ImageLoaderModule {
```
Optional, consider naming this module & the file ``GlideImageLoaderModule`` since it's really providing the Glide implementation of ``ImageLoader``.

veena14cs replied:
renamed to GlideImageLoader.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/ImageLoaderModule.kt:
```diff
@@ -0,0 +1,17 @@
+  @Provides
+  @Singleton
+  @ImageLoaderAnnotation
+  fun providesGlideImageLoader(): ImageLoader {
```
Prefer using ``Binds`` since then you can mark ``GlideImageLoader``'s constructor as ``@Inject``:

```kotlin
@Module
abstract class ImageLoaderModule {
  @Binds
  abstract fun provideGlideImageLoader(impl: GlideImageLoader): ImageLoader
}
```
Note I also removed the ``@Singleton`` scope. We don't need to scope this object because it doesn't contain state, so it's fine if Dagger recreates it each time we inject it.

veena14cs replied:
I tried using @Binds and used above implementation. Its throwing me error
`error: cannot access SimpleTarget class file for com.bumptech.glide.request.target.SimpleTarget not found`

veena14cs replied:
This is removed and resolved error.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/ImageParsingAnnotations.kt:
```diff
@@ -0,0 +1,15 @@
+import javax.inject.Qualifier
+
+/**
+ * These 3 variables provides dependencies for image-extraction URL.
```
Two nits: first, this should be a regular block comment (e.g. ``/* */``) or a single line comment, not a KDoc comment since it's not associated with a specific element. Second, we should probably refer to these as annotations that correspond to dependencies rather than variables (since they aren't actually variables themselves).

veena14cs replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,88 @@
+) : Html.ImageGetter {
+
+
+  /***
```
Nit: ``/** */``, there's an extra ``*`` here.

veena14cs replied:
Removed extra *.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,88 @@
+    internal var drawable: Drawable? = null
+
+    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
+      drawable = BitmapDrawable(context.resources, resource)
```
Let's make this a local variable. I don't see a reason why it needs to be a class variable.

veena14cs replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,88 @@
+    @ImageLoaderAnnotation private val imageLoader: ImageLoader
+  ) {
+    fun create(htmlContentTextView: TextView, entityType: String, entityId: String): UrlImageParser {
+      return UrlImageParser(context,gcsPrefix,gcsResource,imageDownloadUrlTemplate,htmlContentTextView, entityType, entityId,imageLoader)
```
Please put spaces between arguments. If you need to line wrap, you can do it like so:

```kotlin
return UrlImageParser(
  context,
  gcsPrefix,
  gcsResource,
  imageDownloadUrlTemplate,
  htmlContentTextView,
  entityType,
  entityId,
  imageLoader
)
```

veena14cs replied:
Corrected.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/HtmlParser.kt:
```diff
@@ -0,0 +1,51 @@
+
+    val imageGetter =  urlImageParserFactory.create(htmlContentTextView, entityType, entityId)
+    return if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
+      Html.fromHtml(htmlContent, Html.FROM_HTML_MODE_LEGACY, imageGetter, /* tagHandler= */null) as Spannable
```
Nit: ``/* tagHandler= */ null`` (need space between ``*/`` and ``null``).

Ditto below.

veena14cs replied:
Done.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/parser/HtmlParserTest.kt:
```diff
@@ -0,0 +1,84 @@
+  @Test
+  fun testHtmlContent_handleCustomOppiaTags_parsedHtmldisplaysStyledText() {
+    activityTestRule.activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
+    activityTestRule.activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
```
Please remove these two lines--as we discussed via chat, they aren't actually affecting the behavior of the test. Ditto below.

veena14cs replied:
removed orientation lines.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/parser/HtmlParserTest.kt:
```diff
@@ -0,0 +1,84 @@
+    val textView = activityTestRule.activity.findViewById(R.id.test_html_content_text_view) as TextView
+    val htmlResult: Spannable = htmlParserFactory.create( /* entityType= */ "",  /* entityId= */ "")
+      .parseOppiaHtml(
+      "\u003cp\u003e\"Let's try one last question,\" said Mr. Baker. \"Here's a pineapple cake cut into pieces.\"\u003c/p\u003e\u003coppia-noninteractive-image alt-with-value=\"\u0026amp;quot;Pineapple cake with 7/9 having cherries.\u0026amp;quot;\" caption-with-value=\"\u0026amp;quot;\u0026amp;quot;\" filepath-with-value=\"\u0026amp;quot;pineapple_cake_height_479_width_480.png\u0026amp;quot;\"\u ...
```
I think this wrapping is off. When wrapping elements for function calls, the arguments should be indented. I think in this case you're cascading a few different things, and this can probably be simplified a bit by extracting the HTML parser to its own variable:

```kotlin
val htmlParser = htmlParserFactory.create(/* entityType= */ "", /* entityId= */ "")
val htmlResult = htmlParser.parseOppiaHtml(
  "\u003..."
  + "..."
  + "...",
  textView
)
```

Also, please break the string up into new lines rather than having a really long line. You can just line break the string by pressing 'enter' in Android Studio and it will automatically close the string and continue them on the following line.

veena14cs replied:
Corrected.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/parser/HtmlParserTest.kt:
```diff
@@ -0,0 +1,84 @@
+
+/** Tests for [HtmlParser]. */
+@RunWith(AndroidJUnit4::class)
+class HtmlParserTest {
```
Please add a test to verify that image loading works. If you swap out the ImageLoader for a mock of it, you should be able to verify that it gets called.

veena14cs replied:
@BenHenning I have added testcases but it fails I have tried this approach [app/src/sharedTest/java/org/oppia/app/parser/UrlImageParserTest.kt.](https://github.com/oppia/oppia-android/blob/385b1d1d60131970f93dacc9dc91790d81e33c35/app/src/sharedTest/java/org/oppia/app/parser/UrlImageParserTest.kt)

I tried testing with `place_holders` but I cant add place_holder in `utility` module so here I am testing if drawable returned is null or not.

BenHenning replied:
Can you please be specific on what is failing? Without more details I can't help debug.

Also when glancing at that test I noticed that you're replacing the injected image loader with a mock. We need to bind a mock of ImageLoader to the Dagger graph instead of pulling in the Glide version (we shouldn't use the GlideImageLoader in tests) in order for the prod code to call into the mock that the test is setting up.

veena14cs replied:
It fails to load image. It is returning null for drawable.

I tried changing implementation using bind, but it was throwing me error https://github.com/oppia/oppia-android/pull/205#discussion_r337332902

BenHenning replied:
As discussed via Chat and email, this isn't possible to test without fully resolving #59.

veena14cs replied:
Filed issue for this on #277 .

BenHenning reviewed (changes requested):
Thanks @veena14cs! This generally LGTM, but I'd like to see a test that calls through the image loading code now that you set up the ImageLoader interface to help with this.

veena14cs committed:
Merge branch 'develop' into introduce-image-parser-replica

veena14cs committed:
Update GlideImageLoader.kt

veena14cs committed:
removed annotation

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
line break

veena14cs committed:
fix

veena14cs committed:
fix comments

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
injected context

veena14cs committed:
working on testcase.

veena14cs committed:
testcase added

veena14cs committed:
added test case

veena14cs committed:
testcases

veena14cs committed:
Update build.gradle

veena14cs committed:
Update AndroidManifest.xml

veena14cs committed:
fixes

rt4914 commented:
Other than test-cases everything is fine. I am not sure whether the approach for image-testing is correct or not. @BenHenning needs to verify that.

BenHenning commented:
Please see my reply to the open question about the image load testing and follow-up, then assign back to me.

veena14cs committed:
Update UrlImageParserTest.kt

veena14cs committed:
changed to CutomTarget

veena14cs committed:
Update UrlImageParserTest.kt

veena14cs committed:
updated.

BenHenning commented:
@veena14cs please update this with the latest ``develop`` and resolve existing conflicts, then send back to me for review.

veena14cs committed:
Update UrlImageParserTest.kt

veena14cs committed:
Merge branch 'develop' into introduce-image-parser-replica

veena14cs committed:
Update HtmlParserTestActivity.kt

veena14cs commented:
@BenHenning this PR is up-to-date with develop. PTAL

veena14cs committed:
Update misc.xml

veena14cs committed:
Update GlideImageLoaderModule.kt

BenHenning commented on app/src/sharedTest/java/org/oppia/app/parser/HtmlParserTest.kt:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  @Test
+  fun testHtmlContent_handleCustomOppiaTags_parsedHtmldisplaysStyledText() {
```
Nit: ``testHtmlContent_handleCustomOppiaTags_parsedHtmlDisplaysStyledText``

Please fix the typo in this test name,

veena14cs replied:
Done.

BenHenning reviewed (approved):
Thanks @veena14cs. Please remove the mock ImageLoader pieces and failing test for the ImageParser, and add a TODO on #277 to add tests for the class before submitting this.

veena14cs committed:
Update HtmlParserTest.kt

veena14cs committed:
deleted test class for loading image and added todo

veena14cs committed:
updated

veena14cs commented:
> Thanks @veena14cs. Please remove the mock ImageLoader pieces and failing test for the ImageParser, and add a TODO on #277 to add tests for the class before submitting this.

Removed  mock ImageLoader  and added Todo in UrlImageParser class.

veena14cs committed:
Update build.gradle

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -0,0 +1,93 @@
+import com.bumptech.glide.request.transition.Transition
+import javax.inject.Inject
+
+// TODO(#169,#277): Replace this with exploration asset downloader and add test cases for loading image
```
Let's frame these as 2 TODOs rather than combining them since they are separate tasks to complete.

veena14cs replied:
okay. done.

BenHenning reviewed (approved):
Thanks!

veena14cs committed:
Update GlideImageLoader.kt

veena14cs committed:
Update build.gradle

veena14cs committed:
Update UrlImageParser.kt
